# What you can learn about food by analyzing a million Yelp reviews

## The Yelp Dataset
[**The Yelp Dataset**](https://www.yelp.com/dataset_challenge/) is a dataset published by the business review service [Yelp](http://yelp.com) for academic research and educational purposes. I really like the Yelp dataset as a subject for machine learning and natural language processing demos, because it's big (but not so big that you need your own data center to process it), well-connected, and anyone can relate to it &mdash; it's largely about food, after all!

**Note:** If you'd like to execute this notebook interactively on your local machine, you'll need to download your own copy of the Yelp dataset. If you're reviewing a static copy of the notebook online, you can skip this step. Here's how to get the dataset:
1. Please visit the Yelp dataset webpage [here](https://www.yelp.com/dataset_challenge/)
1. Click "Get the Data"
1. Please review, agree to, and respect Yelp's terms of use!
1. The dataset downloads as a compressed .tgz file; uncompress it
1. Place the uncompressed dataset files (*yelp_academic_dataset_business.json*, etc.) in a directory named *yelp_dataset_challenge_academic_dataset*
1. Place the *yelp_dataset_challenge_academic_dataset* within the *data* directory in the *Modern NLP in Python* project folder

That's it! You're ready to go.

The current iteration of the Yelp dataset (as of this demo) consists of the following data:
- __552K__ users
- __77K__ businesses
- __2.2M__ user reviews

When focusing on restaurants alone, there are approximately __22K__ restaurants with approximately __1M__ user reviews written about them.

The data is provided in a handful of files in _.json_ format. We'll be using the following files for our demo:
- __yelp\_academic\_dataset\_business.json__ &mdash; _the records for individual businesses_
- __yelp\_academic\_dataset\_review.json__ &mdash; _the records for reviews users wrote about businesses_

The files are text files (UTF-8) with one _json object_ per line, each one corresponding to an individual data record. Let's take a look at a few examples.

In [6]:
import os
import re
import json
data_directory = os.path.join('..', 'data',
                              'yelp_dataset_challenge_academic_dataset', 'dataset')

The review records are stored in a similar manner &mdash; _key, value_ pairs containing information about the reviews.

In [3]:
review_filepath = os.path.join(data_directory,
                                    'review.json')

with open(review_filepath, encoding='utf_8') as f:
    first_review_record = f.readline()
    
print(first_review_record)

{"review_id":"VfBHSwC5Vz_pbFluy07i9Q","user_id":"cjpdDjZyprfyDG3RlkVG3w","business_id":"uYHaNptLzDLoV_JZ_MuzUA","stars":5,"date":"2016-07-12","text":"My girlfriend and I stayed here for 3 nights and loved it. The location of this hotel and very decent price makes this an amazing deal. When you walk out the front door Scott Monument and Princes street are right in front of you, Edinburgh Castle and the Royal Mile is a 2 minute walk via a close right around the corner, and there are so many hidden gems nearby including Calton Hill and the newly opened Arches that made this location incredible.\n\nThe hotel itself was also very nice with a reasonably priced bar, very considerate staff, and small but comfortable rooms with excellent bathrooms and showers. Only two minor complaints are no telephones in room for room service (not a huge deal for us) and no AC in the room, but they have huge windows which can be fully opened. The staff were incredible though, letting us borrow umbrellas for t

A few attributes of note on the review records:
- __text__ &mdash; _the natural language text the user wrote_
- __stars__ &mdash; _the number of stars the reviewer left_

The _text_ and the _stars_ attribute will be our focus today!

Next, we will create a new directory that contains only the text from reviews about restaurants, with one review per line in the file.

In [23]:
intermediate_directory = os.path.join('..','data','yelp_dataset_challenge_academic_dataset', 'extracted_from_json')

review_txt_filepath = os.path.join(intermediate_directory,'sentiment_data',
                                   'review_text_for_sentiment.txt')

review_sentiment_filepath = os.path.join(intermediate_directory, 'sentiment_data','sentiment_of_review_text.txt')

review_json_filepath = os.path.join(data_directory,'review.json')

In [24]:
%%time
# Make the if statement True
# if you want to execute data prep yourself.

if True:
    
    review_count = 0

    # create & open a new files in write mode
    with open(review_txt_filepath, 'w', encoding='utf_8') as review_txt_file:
        with open(review_sentiment_filepath, 'w', encoding='utf_8') as review_sentiment_file:

            # open the existing review json file
            with open(review_json_filepath, encoding='utf_8') as review_json_file:
                # loop through all reviews in the existing file and convert to dict
                for review_json in review_json_file:
                    review = json.loads(review_json)
                    # write the review as a line in the new file
                    # escape newline characters in the original review text
                    review_txt_file.write(review.get('text','NA').replace('\n', '\\n') + '\n')
                    review_sentiment_file.write(str(review.get('stars','NA')) +'\n')
                    review_count =  review_count + 1

    print ('Text from {} reviews written to the new txt file.'.format(review_count))
    
else:
    
    with open(review_txt_filepath, encoding='utf_8') as review_txt_file:
        for review_count, line in enumerate(review_txt_file):
            pass
        
    print('Text from {} reviews in the txt file.'.format(review_count + 1))



Text from 4736897 reviews written to the new txt file.
CPU times: user 1min 7s, sys: 18 s, total: 1min 25s
Wall time: 1min 28s


In [25]:
#count the lines in the above files

from itertools import (takewhile,repeat)

def rawincount(filename):
    with open(filename, 'rb') as f:
        bufgen = takewhile(lambda x: x, (f.raw.read(1024*1024) for _ in repeat(None)))
        return sum( buf.count(b'\n') for buf in bufgen )

print('Len of review text file:{}\nLen of review sentiment file:{}'.format(rawincount(review_txt_filepath), rawincount(review_sentiment_filepath)))

Len of review text file:4736897
Len of review sentiment file:4736897


In [ ]:
# Good!  The lengths of the files match!

## spaCy &mdash; Industrial-Strength NLP in Python

![spaCy](https://s3.amazonaws.com/skipgram-images/spaCy.png)

[**spaCy**](https://spacy.io) is an industrial-strength natural language processing (_NLP_) library for Python. spaCy's goal is to take recent advancements in natural language processing out of research papers and put them in the hands of users to build production software.

spaCy handles many tasks commonly associated with building an end-to-end natural language processing pipeline:
- Tokenization
- Text normalization, such as lowercasing, stemming/lemmatization
- Part-of-speech tagging
- Syntactic dependency parsing
- Sentence boundary detection
- Named entity recognition and annotation

In the "batteries included" Python tradition, spaCy contains built-in data and models which you can use out-of-the-box for processing general-purpose English language text:
- Large English vocabulary, including stopword lists
- Token "probabilities"
- Word vectors

spaCy is written in optimized Cython, which means it's _fast_. According to a few independent sources, it's the fastest syntactic parser available in any language. Key pieces of the spaCy parsing pipeline are written in pure C, enabling efficient multithreading (i.e., spaCy can release the _GIL_).

In [19]:
#!python -m spacy download en_core_web_md
#!python -m spacy link en_core_web_md en_default


    Link en_default already exists

    To overwrite an existing link, use the --force flag.



In [20]:
import spacy
import pandas as pd
import itertools as it
nlp = spacy.load('en_default')

Let's grab a sample review to play with.

In [21]:
with open(review_txt_filepath, encoding='utf_8') as f:
    sample_review = list(it.islice(f, 8, 9))[0]
    sample_review = sample_review.replace('\\n', '\n')
        
print(sample_review)

First time at this group of hotels. Pretty new, only one in UK, another to open in Edinburgh and one in London. Rooms not very big but great price and location for a weekend in Edinburgh. Rooms clean, comfortable, good shower and free wifi!



## Word Vector Embedding with Word2Vec

Pop quiz! Can you complete this text snippet?

<br><br>

![word2vec quiz](https://s3.amazonaws.com/skipgram-images/word2vec-1.png)

<br><br><br>
You just demonstrated the core machine learning concept behind word vector embedding models!
<br><br><br>

![word2vec quiz 2](https://s3.amazonaws.com/skipgram-images/word2vec-2.png)

The goal of *word vector embedding models*, or *word vector models* for short, is to learn dense, numerical vector representations for each term in a corpus vocabulary. If the model is successful, the vectors it learns about each term should encode some information about the *meaning* or *concept* the term represents, and the relationship between it and other terms in the vocabulary. Word vector models are also fully unsupervised &mdash; they learn all of these meanings and relationships solely by analyzing the text of the corpus, without any advance knowledge provided.

Perhaps the best-known word vector model is [word2vec](https://arxiv.org/pdf/1301.3781v3.pdf), originally proposed in 2013. The general idea of word2vec is, for a given *focus word*, to use the *context* of the word &mdash; i.e., the other words immediately before and after it &mdash; to provide hints about what the focus word might mean. To do this, word2vec uses a *sliding window* technique, where it considers snippets of text only a few tokens long at a time.

At the start of the learning process, the model initializes random vectors for all terms in the corpus vocabulary. The model then slides the window across every snippet of text in the corpus, with each word taking turns as the focus word. Each time the model considers a new snippet, it tries to learn some information about the focus word based on the surrouding context, and it "nudges" the words' vector representations accordingly. One complete pass sliding the window across all of the corpus text is known as a training *epoch*. It's common to train a word2vec model for multiple passes/epochs over the corpus. Over time, the model rearranges the terms' vector representations such that terms that frequently appear in similar contexts have vector representations that are *close* to each other in vector space.

For a deeper dive into word2vec's machine learning process, see [here](https://arxiv.org/pdf/1411.2738v4.pdf).

Word2vec has a number of user-defined hyperparameters, including:
- The dimensionality of the vectors. Typical choices include a few dozen to several hundred.
- The width of the sliding window, in tokens. Five is a common default choice, but narrower and wider windows are possible.
- The number of training epochs.

For using word2vec in Python, [gensim](https://rare-technologies.com/deep-learning-with-word2vec-and-gensim/) comes to the rescue again! It offers a [highly-optimized](https://rare-technologies.com/word2vec-in-python-part-two-optimizing/), [parallelized](https://rare-technologies.com/parallelizing-word2vec-in-python/) implementation of the word2vec algorithm with its [Word2Vec](https://radimrehurek.com/gensim/models/word2vec.html) class.

In [58]:
from gensim.models import Word2Vec

trigram_sentences = LineSentence(trigram_sentences_filepath)
word2vec_filepath = os.path.join(intermediate_directory, 'word2vec_model_all')

We'll train our word2vec model using the normalized sentences with our phrase models applied. We'll use 100-dimensional vectors, and set up our training process to run for twelve epochs.

In [59]:
%%time

# this is a bit time consuming - make the if statement True
# if you want to train the word2vec model yourself.
if 0 == 1:

    # initiate the model and perform the first epoch of training
    food2vec = Word2Vec(trigram_sentences, size=100, window=5,
                        min_count=20, sg=1, workers=4)
    
    food2vec.save(word2vec_filepath)

    # perform another 11 epochs of training
    for i in range(1,12):

        food2vec.train(trigram_sentences)
        food2vec.save(word2vec_filepath)
        
# load the finished model from disk
food2vec = Word2Vec.load(word2vec_filepath)
food2vec.init_sims()

print u'{} training epochs so far.'.format(food2vec.train_count)

12 training epochs so far.
CPU times: user 5.43 s, sys: 891 ms, total: 6.32 s
Wall time: 7.12 s


On my four-core machine, each epoch over all the text in the ~1 million Yelp reviews takes about 5-10 minutes.

In [60]:
print u'{:,} terms in the food2vec vocabulary.'.format(len(food2vec.vocab))

50,835 terms in the food2vec vocabulary.


Let's take a peek at the word vectors our model has learned. We'll create a pandas DataFrame with the terms as the row labels, and the 100 dimensions of the word vector model as the columns.

In [90]:
# build a list of the terms, integer indices,
# and term counts from the food2vec model vocabulary
ordered_vocab = [(term, voc.index, voc.count)
                 for term, voc in food2vec.vocab.iteritems()]

# sort by the term counts, so the most common terms appear first
ordered_vocab = sorted(ordered_vocab, key=lambda (term, index, count): -count)

# unzip the terms, integer indices, and counts into separate lists
ordered_terms, term_indices, term_counts = zip(*ordered_vocab)

# create a DataFrame with the food2vec vectors as data,
# and the terms as row labels
word_vectors = pd.DataFrame(food2vec.syn0norm[term_indices, :],
                            index=ordered_terms)

word_vectors

0         1         2         3         4   \
the                      -0.035762 -0.173890 -0.035782 -0.007144  0.032371   
be                       -0.074780 -0.049524  0.085974 -0.098892  0.141556   
and                      -0.070505 -0.026918  0.028344 -0.099909  0.127974   
i                        -0.161238  0.050831 -0.081706 -0.084479  0.053073   
a                        -0.083491 -0.033712 -0.124125 -0.110776 -0.033046   
to                       -0.012082  0.033135 -0.063183 -0.057252 -0.018721   
it                        0.025022  0.081581  0.127987 -0.188015  0.041450   
have                     -0.140812 -0.070552  0.022102  0.001077  0.109890   
of                       -0.036341 -0.054903  0.000644 -0.010602  0.168195   
not                      -0.075276  0.109047  0.055135  0.052251  0.209437   
for                      -0.102976  0.001369 -0.069402 -0.122936  0.028278   
in                       -0.053390 -0.175599 -0.091688 -0.153791  0.003205   
we                       -0.015929 -0.019187 -0.186680 -0.240963  0.077926   
that                     -0.026609  0.085940  0.118164  0.011576  0.156952   
but                      -0.063436  0.089140 -0.057425 -0.093110  0.066531   
with                      0.042318 -0.186670 -0.230563  0.076302  0.216593   
my                       -0.002067 -0.023159  0.035879  0.036316 -0.110738   
this                      0.139739  0.184600  0.137359 -0.109916  0.021484   
you                      -0.171262 -0.119866  0.063801 -0.087287 -0.061923   
on                        0.057481  0.044937 -0.063766 -0.007839  0.161119   
they                     -0.235321 -0.026314  0.143165 -0.170460  0.042189   
food                     -0.164133  0.007745  0.058311 -0.169839 -0.042278   
do                        0.091428 -0.132115  0.105080  0.135949  0.038100   
good                     -0.239592 -0.232940 -0.005036 -0.028226  0.149816   
place                     0.025479  0.130311  0.119834 -0.096365  0.013793   
so                        0.021455  0.079794  0.192058 -0.093809 -0.094279   
get                       0.009313 -0.101684 -0.163864 -0.159002  0.018936   
go                        0.031094 -0.126839 -0.054429 -0.221885 -0.063464   
at                        0.102501 -0.095756 -0.216304 -0.107230 -0.112544   
as                        0.141091  0.073669  0.109637 -0.112564 -0.167600   
...                            ...       ...       ...       ...       ...   
hard_boiled_egg           0.025154  0.060949 -0.064816  0.071975  0.087870   
poached_quail_egg         0.060616  0.001519 -0.052908  0.013590  0.031732   
egg_foo                   0.022816  0.077288 -0.122898 -0.022765 -0.137979   
sherri                   -0.089536 -0.030370  0.011311 -0.094766 -0.121970   
hindrance                -0.033325  0.011520  0.027370  0.240223 -0.004098   
eggdrop_soup             -0.028369  0.040039 -0.005378 -0.088736  0.015203   
arbitrarily               0.203714 -0.047405 -0.045261  0.000302 -0.074105   
faisant                   0.046053  0.083873  0.057943  0.174203 -0.121259   
marian                   -0.035330  0.146843 -0.173594 -0.010971 -0.150150   
9:30p                     0.119010  0.002159 -0.000148 -0.102635 -0.016918   
extra_chasu              -0.225889 -0.131615  0.046431  0.017999  0.119188   
lavosh_wrap               0.134420 -0.032055  0.012240  0.024420  0.031334   
dum_dum                  -0.002741 -0.137371  0.030704 -0.030365 -0.134645   
triplet                  -0.010495  0.057432 -0.019535 -0.044881  0.042409   
nantucket                -0.124356  0.141918 -0.038579  0.035650 -0.157662   
gurl                     -0.119794  0.025898 -0.070130 -0.027929  0.113244   
nordstroms               -0.030410 -0.026861 -0.016836  0.097363 -0.098189   
eau_de                   -0.146450  0.000788 -0.094391  0.146833 -0.051392   
extra_.5                 -0.124972 -0.032596 -0.017800  0.106415 -0.086728   
hazelnut_crunch           0.055723 -0.098708  0.013225

Holy wall of numbers! This DataFrame has 50,835 rows &mdash; one for each term in the vocabulary &mdash; and 100 colums. Our model has learned a quantitative vector representation for each term, as expected.

Put another way, our model has "embedded" the terms into a 100-dimensional vector space.

### So... what can we do with all these numbers?
The first thing we can use them for is to simply look up related words and phrases for a given term of interest.

In [63]:
def get_related_terms(token, topn=10):
    """
    look up the topn most similar terms to token
    and print them as a formatted list
    """

    for word, similarity in food2vec.most_similar(positive=[token], topn=topn):

        print u'{:20} {}'.format(word, round(similarity, 3))